src/compiler/checker.ts

```diff
@@ -2104,7 +2104,17 @@ namespace ts {
             if (ambientModule) {
                 return ambientModule;
             }
-            const resolvedModule = getResolvedModule(getSourceFileOfNode(location), moduleReference);
+            const currentSourceFile = getSourceFileOfNode(location);
+            const resolvedModuleState = getResolvedModule(currentSourceFile, moduleReference);
+            let resolvedModule: ResolvedModuleFull;
+            if (currentSourceFile && resolvedModuleState === undefined) {

```

Comment:
not sure i understand why we need to do the resolution on two phases, given that we only support string literals as arguments to `import`

Reply:
> given that we only support string literals as arguments to import

String literal _types_. Right now it's not an error to say

```ts
type Alias = "module";
const x: import(Alias);
```

the modules in question just only get loaded if they're already included in the build, otherwise produce an error. (see `importTypeNested` test.) I can restrain this further, but I don't much see the point in limiting it? I could also handle unions pretty easily as it turns out, too (by bypassing normal typequery logic I've given myself a place to return union types and do multiple namespace lookups).

Reply:
As we have chatted offline. i think the real value of such feature is to enable higher order types, since we have decided not to do that, i would say no need for the complexity. I would just remove all the deferred module resolution stuff, and keep it simple.

Reply:
parsing a type to be future proof is fine. but just make it such that the argument has to be string literal node.

Reply:
@mhegazy Done. It now issues an error if not a string literal, and the late-resolution code has been removed.